src/VecSim/algorithms/brute_force/bf_batch_iterator.cpp

```diff
@@ -0,0 +1,139 @@
+#include <cassert>
+#include <algorithm>
+#include <iostream>
+#include <cmath>
+#include <functional>
+#include "bf_batch_iterator.h"
+#include "VecSim/utils/arr_cpp.h"
+#include "VecSim/utils/vec_utils.h"
+#include "VecSim/query_result_struct.h"
+
+unsigned char BF_BatchIterator::next_id = 0;
+
+// Compare function that handle NaN values (consider these as the highest, above inf)
+bool cmp(const pair<float, labelType> &a, const pair<float, labelType> &b) {
+    if (isnan(a.first)) {
+        return false;
+    }
+    if (isnan(b.first)) {
+        return true;
+    }
+    return a.first < b.first;
+}
+
+// heuristics: decide if using heap or select search, based on the ratio between the
+// number of remaining results and the index size.
+VecSimQueryResult *BF_BatchIterator::searchByHeuristics(size_t n_res,
+                                                        VecSimQueryResult_Order order) {
+    if ((getIndex()->indexSize() - getResultsCount()) / 1000 > n_res) {
+        return heapBasedSearch(n_res);
```

Comment:
you are missing ordering here

Reply:
What do you mean? the heap based search always returns the results sorted by their scores.